tests/providers/microsoft/azure/triggers/test_powerbi.py

```diff
@@ -91,167 +81,170 @@ def test_powerbi_trigger_serialization():
+        mock_get_refresh_details_by_refresh_id.side_effect = Exception("Test exception")
+        mock_cancel_dataset_refresh.side_effect = Exception("Exception caused by cancel_dataset_refresh")
+        mock_trigger_dataset_refresh.return_value = DATASET_REFRESH_ID
+
+        task = [i async for i in powerbi_trigger.run()]
+        response = TriggerEvent(
+            {
+                "status": "error",
+                "message": "An error occurred while canceling dataset: Exception caused by cancel_dataset_refresh",
+                "dataset_refresh_id": DATASET_REFRESH_ID,
+            }
+        )
 
-@pytest.mark.asyncio
-@mock.patch(f"{MODULE}.hooks.powerbi.PowerBIHook.get_refresh_details_by_refresh_id")
-@mock.patch(f"{MODULE}.hooks.powerbi.PowerBIHook.trigger_dataset_refresh")
-async def test_powerbi_trigger_run_timeout(
-    mock_trigger_dataset_refresh, mock_get_refresh_details_by_refresh_id, powerbi_trigger
-):
-    """Assert that powerbi run timesout after end_time elapses"""
-    mock_get_refresh_details_by_refresh_id.return_value = {"status": PowerBIDatasetRefreshStatus.IN_PROGRESS}
-    mock_trigger_dataset_refresh.return_value = DATASET_REFRESH_ID
+        assert len(task) == 1
+        assert response in task
+        mock_cancel_dataset_refresh.assert_called_once()
 
-    generator = powerbi_trigger.run()
-    actual = await generator.asend(None)
-    expected = TriggerEvent(
-        {
-            "status": "error",
-            "message": f"Timeout occurred while waiting for dataset refresh to complete: The dataset refresh {DATASET_REFRESH_ID} has status In Progress.",
-            "dataset_refresh_id": DATASET_REFRESH_ID,
+    @pytest.mark.asyncio
+    @mock.patch(f"{MODULE}.hooks.powerbi.PowerBIHook.get_refresh_details_by_refresh_id")
+    @mock.patch(f"{MODULE}.hooks.powerbi.PowerBIHook.trigger_dataset_refresh")
+    async def test_powerbi_trigger_run_exception_without_refresh_id(
+        self, mock_trigger_dataset_refresh, mock_get_refresh_details_by_refresh_id, powerbi_trigger
+    ):
+        """Assert handling of exception when there is no dataset_refresh_id"""
+        powerbi_trigger.dataset_refresh_id = None
+        mock_get_refresh_details_by_refresh_id.side_effect = Exception(
+            "Test exception for no dataset_refresh_id"
+        )
+        mock_trigger_dataset_refresh.return_value = None
+
+        task = [i async for i in powerbi_trigger.run()]
+        response = TriggerEvent(
+            {
+                "status": "error",
+                "message": "An error occurred: Test exception for no dataset_refresh_id",
+                "dataset_refresh_id": None,
+            }
+        )
+        assert len(task) == 1
+        assert response in task
+
+    @pytest.mark.asyncio
+    @mock.patch(f"{MODULE}.hooks.powerbi.PowerBIHook.get_refresh_details_by_refresh_id")
+    @mock.patch(f"{MODULE}.hooks.powerbi.PowerBIHook.trigger_dataset_refresh")
+    async def test_powerbi_trigger_run_timeout(
```

Comment:
Hi @dabla I just checked some tests and realized that this test added in this PR always needs 30 seconds - presumably because of the `TIMEOUT` being defined =30 above.

This is not really acceptable and I propose to use a better method such that the pytest does not run 30 seconds.

Can you please improve this? THANKS.

If you want to test the runtime, you can check selectively via `breeze testing tests --test-type "Providers[microsoft]" -- providers/tests/microsoft/azure/triggers/test_powerbi.py::TestPowerBITrigger::test_powerbi_trigger_run_timeout`

Reply:
I’m willing to look at it but I think this test was originally added by @ambika-garg as this is regarding the PowerBi operator, not the MSGraphOperator. Those tests were probably updated by me due to the change of api version definitions.

Reply:
Hi @jscheffl, Thank you for reviewing it. I will look into this test.

Reply:
> Hi @jscheffl, Thank you for reviewing it. I will look into this test.

Thank you @ambika-garg for looking into it.